Adding mdot to history and moving history to new analysis/ directory

AstroBarker opened the pull request:
I've added the mass accretion rate to Phoebus history output. I also moved `history.hpp` to a new directly, `analysis`. I created `analysis/history_utils.hpp` for calculating the mass flux term and, later, other helper functions as we track more complicated quantities. Below is a quick plot of mdot for a 2D GRMHD torus problem.
![image](https://user-images.githubusercontent.com/39746406/199619278-bfe6ecab-8bfd-4b25-b685-0c309f8c840d.png)


- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by calling `scripts/bash/format.sh`.
- [x] Explain what you did.

Yurlungur commented on src/analysis/history.hpp:
```diff
@@ -71,6 +76,59 @@ Real ReduceOneVar(MeshData<Real> *md, const std::string &varname, int idx = 0) {
   return result;
 }
 
+template <typename Reducer_t>
```
If we're always going to sum, you probably don't need to template on reduction here.

Yurlungur commented on src/analysis/history.hpp:
```diff
@@ -71,6 +76,59 @@ Real ReduceOneVar(MeshData<Real> *md, const std::string &varname, int idx = 0) {
+  auto geom = Geometry::GetCoordinateSystem(md);
+
+  Real result = 0.0;
+  Reducer_t reducer(result);
```
If we always sum you don't need to instantiate `reducer` you can just leave it as `result`. Sum is the default reducer in kokkos.

Yurlungur commented on src/analysis/history.hpp:
```diff
@@ -71,6 +76,59 @@ Real ReduceOneVar(MeshData<Real> *md, const std::string &varname, int idx = 0) {
+                       (xh - coords.x1v(i)) * m) *
+                      dx2 * dx3;
+
+          reducer.join(lresult, -flux);
+        } else {
+          reducer.join(lresult, 0.0);
+        }
```
if you are doing sum automatically you can just do `lresult += -flux` and `lresult += 0`

Yurlungur commented on src/analysis/history.hpp:
```diff
@@ -71,6 +76,59 @@ Real ReduceOneVar(MeshData<Real> *md, const std::string &varname, int idx = 0) {
+          reducer.join(lresult, 0.0);
+        }
+      },
+      reducer);
```
if you are always summing you can just leave this as `result`.

Yurlungur commented on src/analysis/history_utils.hpp:
```diff
@@ -0,0 +1,56 @@
+#include <string>
+#include <vector>
+
+#include "geometry/geometry.hpp"
+#include "geometry/geometry_utils.hpp"
+#include "phoebus_utils/relativity_utils.hpp"
+#include "phoebus_utils/variables.hpp"
+#include <kokkos_abstraction.hpp>
+#include <parthenon/package.hpp>
+#include <utils/error_checking.hpp>
```
It doesn't matter much, but I like to space-separate teh parthenon vs phoebus headers. And also, due to lore, I like to order it:
- stdlib headers
- dependencies such as parthenon
- phoebus

so in this case that would look like:
```suggestion
#include <kokkos_abstraction.hpp>
#include <parthenon/package.hpp>
#include <utils/error_checking.hpp>

#include "geometry/geometry.hpp"
#include "geometry/geometry_utils.hpp"
#include "phoebus_utils/relativity_utils.hpp"
#include "phoebus_utils/variables.hpp"
```

Yurlungur commented on src/geometry/geometry.cpp:
```diff
@@ -58,6 +51,22 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+
+    hst_vars.emplace_back(HistoryOutputVar(HstSum, ReduceAccretionRate, "mdot"));
+    params.Add(parthenon::hist_param_key, hst_vars);
+  }
```
Hmm it occurs to me that something gross about putting this in `geometry::Initialize` is that this machinery is present even if the fluid variables are not. (One could imagine running, e.g., radiation around a black hole, for example.)

One can resolve this (in a kind of hacky way) by also checking the following boolean condition from the input deck
```C++
bool hydro_on = pin->GetBoolean("physics", "hydro");
```

Yurlungur reviewed (changes requested):
Thanks for this! Overall this is great! I did notice, however, that this changeset broke unit tests and I think I see why and what the solution is.

Your reduction uses geometry because it pulls in mdot stuff, but some of the geometry machinery is set at configure time. However, that isn't passed into the unit tests, and so the unit tests are complaining that the geometry you've pulled in isn't available.

There are a few solutions:
1. If you make `ReduceMassAccretionRate` not templated on `Reducer_t`, then you can move the implementation into a separate implementation file, e.g., `analysis/history.cpp`. This is safe as far as relocatable device code, because the actual code called inside the `par_reduce` is still "inline." This is the way to go IMO, as it's simple and also will help reduce compile times by compiling the `ReduceMassAccretionRate` function only once.
2. We could pass the configure time information into the unit tests. This would be pretty straightforward, a small cmake tweak or two.
3. We could try and disentangle the history machinery from the unit tests a bit more---it's getting pulled in from the TOV package, because TOV is used as a unit test for the monopole solver. This would be hardest to do but maybe worth it.

I recommend we do option 1. Also a few more comments below:

Yurlungur commented:
@brryan @bprather also interested in your thoughts on this improvement. :)

AstroBarker commented:
Thanks for the feedback @Yurlungur. I've de-templated `ReduceMassAccretionRate`as suggested and moved it into `analysis/history.cpp`. I also added a check for if hydro is active to avoid trying to record mdot with no fluid evolution.

brryan commented on src/analysis/history.cpp:
```diff
@@ -0,0 +1,69 @@
+                   (2.0 * dx1);
+          auto flux = (CalcMassFlux(pack, geom, prho, pvel_lo, pvel_hi, b, k, j, i) +
+                       (xh - coords.x1v(i)) * m) *
+                      dx2 * dx3;
```
I don't quite follow this equation -- I agree with it when `xh = coords.x1v(i)`, but otherwise I'm confused. Why not just find the two indices `i` and `i+1` that contain `xh` and then weight their `CalcMassFlux` values by `(xh - coords.x1v(i)) / (coords.x1v(i+1) - coords.x1v(i))` and `(coords.x1v(i+1) - xh) / (coords.x1v(i+1) - coords.x1v(i))` respectively?

Yurlungur replied:
I think that's equivalent... This version is an interpolation of that right?

AstroBarker replied:
So I used the linear interpolation for `y(xh) = y0 + (xh - x0) m` for slope `m`, which is equivalent to the weighted average form of linear interpolation. A slight difference perhaps, for the slope I used two neighboring cells `i+1` and `i-1`.

brryan replied:
Yeah youre right sorry, I just wasnt thinking clearly.

brryan commented on src/analysis/history_utils.hpp:
```diff
@@ -0,0 +1,56 @@
+  const Real W = phoebus::GetLorentzFactor(vel, gcov4);
+  const Real ucon = vel[0] - shift[0] * W / lapse;
+
+  return gdet * pack(b, prho, k, j, i) * ucon;
```
Do we actually want `lapse * gdet * pack(b, prho, k, j, i) * ucon;`? I think we want the integral of the mass flux over the spherical shell at `i` which I understand to be `\int \sqrt{-g} \rho u^1 dx2 dx3` (the `dx2 dx3` show up later of course). Whereas here `gdet = DetGamma = \sqrt{-g} / \alpha`.

AstroBarker replied:
Good catch, I missed that `gdet` was `det gamma`. Fixing this now.

brryan reviewed (changes requested):
Nice work @AstroBarker! Just some small queries.

Yurlungur commented on src/analysis/history.hpp:
```diff
@@ -21,6 +21,12 @@
 #include <parthenon/package.hpp>
 #include <utils/error_checking.hpp>
 
+#include "geometry/geometry.hpp"
+#include "geometry/geometry_utils.hpp"
+#include "history_utils.hpp"
+#include "phoebus_utils/relativity_utils.hpp"
+#include "phoebus_utils/variables.hpp"
```
Move as many of these include statements as possible out of the header file and into `analysis/history.hpp`. I suspect you can at least move `#include "geometry/geometry.hpp"` and `i#include "phoebus_utils/relativity_utils.hpp"` into `analysis/history.hpp`

AstroBarker replied:
Done, thanks!

Yurlungur reviewed (changes requested):
Thanks for the fixes! The unit tests are still failing because of the include statements below. We should move as many of them as we can into `history/analysis.cpp`.

In general it's good practice to minimize the number of includes in header files, as those includes get recompiled everywhere, and they also introduce additional inter-dependency between source files, which can easily result in circular dependencies.

Yurlungur commented on CMakeLists.txt:
```diff
@@ -187,6 +187,7 @@ add_subdirectory(src)
 # TODO(JMM): Testing?
 if(PHOEBUS_ENABLE_UNIT_TESTS OR PHOEBUS_ENABLE_INTEGRATION_TESTS OR PHOEBUS_ENABLE_REGRESSION_TESTS)
   message("\nConfiguring tests")
+  enable_testing()
```
Needed to get `ctest` to properly register unit tests for some reason.

Yurlungur commented on tst/CMakeLists.txt:
```diff
@@ -26,6 +26,11 @@ if (PHOEBUS_ENABLE_MPI)
     MPI_PARALLEL)
 endif()
 
+if(PHOEBUS_ENABLE_UNIT_TESTS)
+  message("Building unit tests.")
+  add_subdirectory(unit)
+endif()
```
don't know if this change matters. But it doesn't hurt.

Yurlungur commented on tst/unit/geometry/CMakeLists.txt:
```diff
@@ -39,6 +41,10 @@ add_library(geometry_unit_tests OBJECT
+# Super annoying but it had to be done.
+set(MAX_NUMBER_CONSERVED_VARS 9 CACHE STRING "Maximum number of conserved variables")
+configure_file(../../../src/compile_constants.hpp.in generated/compile_constants.hpp @ONLY)
+
```
These were the relevant changes for making the relevant file get autogenerated.

Yurlungur reviewed (approved):
Tests were still failing, so I gave up and generated the relevant file and hooked it into tests. Also fixed a bizarre error where ctest wasn't properly slurping up tests.

Thanks for all your work on this @AstroBarker and for putting up with our nitpicking. As soon as @brryan approves let's merge this in. :)

Yurlungur commented:
@brryan if you're happy with the PR now can you approve?
